torch/utils/data/datapipes/iter/selecting.py

```diff
@@ -16,24 +16,47 @@ class FilterIterDataPipe(MapIterDataPipe):
         filter_fn: Customized function mapping an element to a boolean.
         fn_args: Positional arguments for `filter_fn`
         fn_kwargs: Keyword arguments for `filter_fn`
+        drop_empty_batches:
+        nesting_level:
     """
     def __init__(self,
                  datapipe: IterDataPipe[T_co],
                  filter_fn: Callable[..., bool],
                  fn_args: Optional[Tuple] = None,
                  fn_kwargs: Optional[Dict] = None,
+                 drop_empty_batches: bool = True,
+                 nesting_level: int = 0,
                  ) -> None:
-        super().__init__(datapipe, fn=filter_fn, fn_args=fn_args, fn_kwargs=fn_kwargs)
+        self.drop_empty_batches = drop_empty_batches
+        super().__init__(datapipe, fn=filter_fn, fn_args=fn_args, fn_kwargs=fn_kwargs, nesting_level=nesting_level)
+
+    def _merge(self, data, mask):
+        result = []
+        for i, b in zip(data, mask):
+            if isinstance(b, list):
+                t = self._merge(i, b)
+                if len(t) > 0 or not self.drop_empty_batches:
+                    result.append(t)
+            else:
+                if b:
+                    result.append(i)
+        return result
 
     def __iter__(self) -> Iterator[T_co]:
         res: bool
         for data in self.datapipe:
-            res = self.fn(data, *self.args, **self.kwargs)
-            if not isinstance(res, bool):
-                raise ValueError("Boolean output is required for "
-                                 "`filter_fn` of FilterIterDataPipe")
-            if res:
-                yield data
+            if self.nesting_level == 0 or not isinstance(data, list):
```

Comment:
I have a little bit concern with the logic here. When input data is not a list but `nesting_level` is larger than 0, should we raise `IndexError`? In this case, user must explicitly set the `nesting_level` to a value larger than 0, so the expected behavior is to filter over some certain depth. But, this logic provides users to filter on the data itself.

Reply:
I agree. I am adding an additional check `self.nesting_level > 0 and not isinstance(data, list)` and raise `IndexError` if that condition is met.